Impro 1161 correct coordinate name and units

Katie-Howard opened the pull request:
Addresses #GitHubissuenum

Description

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

CLA
 - [ ] If a new developer, signed up to CLA

cgsandford commented on lib/improver/percentile.py:
```diff
@@ -119,7 +119,19 @@ def process(self, cube):
+                name = name + '_' + self.collapse_coord[i]
+                i += 1
+        else:
+            name = name + '_' + self.collapse_coord
```
As I understand it, you're doing this to construct the name of the percentile coordinate that's created by `cube.collapsed`, so that you know which one to rename?

Can I suggest an alternative using the `improver.utilities.cube_checker.find_percentile_coord` function in the block below:
```
percentile_coord = find_percentile_coord(result)
result.coord(percentile_coord).rename('percentile')
```
This gets rid of the `if / else` statement completely.  Would that work?

Katie-Howard replied:
Yes that is why the if/else statement is there. I didn't know find_percentile_coordinate existed, otherwise I would have used that originally! This is all changed now though.

fionaRust commented:
This looks fine, but when I did a grep for "percentile_over" we still have a lot of references to coordinates of this form. Most of them are in the tests which should be easy to fix, and Caroline pointed out another on the ticket. The two that concern be most are the ones in load and threshold, where they are doing enforcing coordinate order based on "percentile_over" coordinates. These should probably be updated to "percentile" because then I think they will match coordinates in "percentile_over" or just "percentile" form.

Katie-Howard commented:
I have now removed further references to "percentile_over_" within:

- lib/improver/tests/set_up_test_cubes.py

- lib/improver/tests/test_set_up_test_cubes.py

- lib/improver/tests/utilities/test_load.py

- lib/improver/threshold.py

- lib/improver/utilities/load.py

These are the main references that @fionaRust  and @cgsandford mentioned.

fionaRust commented:
Thanks for your latest changes. If we want to tidy up more, this is the result of the grep I've done:
```
lib/improver/tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:86:            'percentile_over_neighbourhood'), iris.coords.Coord)
lib/improver/tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:88:            'percentile_over_neighbourhood').points, DEFAULT_PERCENTILES)
lib/improver/tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:103:            result.coord_dims("percentile_over_neighbourhood")[0], 0)
lib/improver/tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:115:            result.coord_dims("percentile_over_neighbourhood")[0], 0)
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:82:                          long_name="percentile_over_realization")
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:247:                                        long_name="percentile_over_time"))
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:259:                                        long_name="percentile_over_time"), 0)
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:406:        perc_coord = perc_cube.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:429:        perc_coord = perc_cube.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:448:        perc_coord = perc_cube.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:547:        perc_coord = perc_cube.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:565:        perc_coord = perc_cube.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:581:        perc_coord = perc_cube.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:602:        pl_perc_coord = percentile_leading.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:603:        tl_perc_coord = time_leading.coord('percentile_over_realization')
lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py:628:        perc_coord = longitude_leading.coord('percentile_over_realization')
lib/improver/tests/utilities/test_cube_checker.py:419:                perc_name='percentile_over_dummy',
lib/improver/tests/utilities/test_cube_checker.py:426:        self.assertEqual(perc_coord.name(), "percentile_over_dummy")
lib/improver/tests/utilities/test_cube_checker.py:440:        cube.remove_coord("percentile_over_dummy")
lib/improver/tests/utilities/test_cube_checker.py:450:                                 long_name='percentile_over_realization',
lib/improver/tests/wind_calculations/wind_gust_diagnostic/test_WindGustDiagnostic.py:50:        perc_name='percentile_over_neighbourhood',
lib/improver/tests/wind_calculations/wind_gust_diagnostic/test_WindGustDiagnostic.py:157:                                   long_name="percentile_over_neighbourhood",
lib/improver/tests/wind_calculations/wind_gust_diagnostic/test_WindGustDiagnostic.py:244:        self.assertEqual(perc_coord.name(), "percentile_over_neighbourhood")
lib/improver/tests/wind_calculations/wind_gust_diagnostic/test_WindGustDiagnostic.py:245:        self.assertEqual(result.coord("percentile_over_neighbourhood").points,
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py:71:        self.perc_coord = "percentile_over_neighbourhood"
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py:582:        self.perc_coord = "percentile_over_neighbourhood"
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py:607:        cube.coord(self.perc_coord).rename("percentile_over_dummy")
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_GeneratePercentilesFromMeanAndVariance.py:361:            len(result.coord("percentile_over_realization").points))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_GeneratePercentilesFromMeanAndVariance.py:397:            len(result.coord("percentile_over_realization").points))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_GeneratePercentilesFromMeanAndVariance.py:399:            percentiles, result.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsMembers.py:63:        cube.coord("realization").rename("percentile_over_realization")
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsMembers.py:80:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsMembers.py:92:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsMembers.py:102:        cube.coord("percentile_over_realization").rename(
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsMembers.py:103:            "percentile_over_neighbourhood")
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsMembers.py:104:        plen = len(cube.coord("percentile_over_neighbourhood").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:64:        cube.coord("realization").rename("percentile_over_realization")
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:65:        cube.coord("percentile_over_realization").points = (
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:73:        percentiles = cube.coord("percentile_over_realization").points
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:88:        percentiles = cube.coord("percentile_over_realization").points
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:104:        percentiles = cube.coord("percentile_over_realization").points
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:200:        self.perc_coord = "percentile_over_neighbourhood"
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:490:        self.perc_coord = "percentile_over_realization"
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:497:        self.perc_coord_mismatch = "percentile_over_neighbourhood"
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:249:            result.coord("percentile_over_realization"), DimCoord)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:251:            result.coord("percentile_over_realization").points, percentiles)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:266:            result.coord("percentile_over_realization"), DimCoord)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:268:            result.coord("percentile_over_realization").points, percentiles)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:434:        cube.coord("realization").rename("percentile_over_realization")
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:442:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:445:                cube.data, cube, "percentile_over_realization", plen))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:455:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:458:                cube.data, cube, "percentile_over_realization", plen))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:471:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:475:                cube.data, cube, "percentile_over_realization", plen)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:488:        for cube_slice in cube.slices_over("percentile_over_realization"):
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:490:        plen = len(cube_slice.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:494:                "percentile_over_realization", plen))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:517:        cube.coord("realization").rename("percentile_over_realization")
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:518:        cube.coord("percentile_over_realization").points = (
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:528:                "percentile_over_realization", plen))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:538:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:541:                flattened_data, cube, "percentile_over_realization", plen))
lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ensemble_copula_coupling_utilities.py:552:        plen = len(cube.coord("percentile_over_realization").points)
lib/improver/nbhood/circular_kernel.py:451:        if result.coords("percentile_over_neighbourhood"):
lib/improver/nbhood/circular_kernel.py:453:                result.coord_dims("percentile_over_neighbourhood")[0])
lib/improver/nbhood/circular_kernel.py:457:                                    "percentile_over_neighbourhood"]:
lib/improver/nbhood/circular_kernel.py:477:        pct_coord_name = "percentile_over_neighbourhood"
```
All the changes in the unit tests should be easy to fix. I'm not sure about the percentile_over_neighbourhood cases, we should probably check with Caroline and/or Bruce

Katie-Howard commented:
I have removed the references to "percentile_over_" from all the files above and replaced with "percentiles". When running the CLI tests, this did cause two of them to fail, but this is because the output files had references to "percentile_over" within them. I have regenerated this KGO and they all now run fine.

cgsandford commented on lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py:
```diff
@@ -604,7 +604,7 @@ def test_basic(self):
     def test_works_for_different_percentile_coord(self):
         """Test that it still works for different percentile coordinate"""
         cube = self.post_processed_percentiles
-        cube.coord(self.perc_coord).rename("percentile_over_dummy")
```
Could probably get rid of this test now that there is only one "percentile" coordinate name.

Katie-Howard replied:
Done

cgsandford commented on lib/improver/tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:
```diff
@@ -487,14 +487,14 @@ def setUp(self):
         self.percentile_cube = (
             add_forecast_reference_time_and_forecast_period(cube))
-        self.perc_coord_mismatch = "percentile_over_neighbourhood"
+        self.perc_coord_mismatch = "percentile"
```
Looking at the context around this test.  The setup function with the same named percentile coordinate now sets up two identical cubes (ie `self.percentile_cube` = `self.percentile_cube_mismatch`).  This and the associated test at line 552 (`test_ok_for_diff_percentile_coord`) are no longer needed and can be removed.

Katie-Howard replied:
Done

cgsandford replied:
Thanks.

cgsandford commented on lib/improver/tests/set_up_test_cubes.py:
```diff
@@ -257,7 +257,7 @@ def set_up_variable_cube(data, name='air_temperature', units='K',
 def set_up_percentile_cube(data, percentiles, name='air_temperature',
                            units='K', spatial_grid='latlon',
-                           percentile_dim_name='percentile_over_realization',
+                           percentile_dim_name='percentile',
```
I'd actually be in favour of removing this option completely, but that might require more involved changes and this PR is getting complex enough!  I'll note it on the ticket and I or somebody else can do it when we get around to it.

Katie-Howard replied:
Yes it might be best done in a different PR now.

cgsandford commented on lib/improver/tests/utilities/test_cube_checker.py:
```diff
@@ -447,7 +447,7 @@ def test_fails_if_too_many_perc_coord(self):
         new_perc_coord = (
             iris.coords.AuxCoord(1,
-                                 long_name='percentile_over_realization',
+                                 long_name='percentile',
```
Does this still work?  I'm surprised it lets you add a duplicate coordinate.

Katie-Howard replied:
Yes all the unit tests still work, I'll check this test again though.

cgsandford replied:
No, I'm just surprised.  Maybe the data are different.  But this is probably my misunderstanding iris so I'll poke at this code in my own time.

cgsandford commented on lib/improver/tests/utilities/test_load.py:
```diff
@@ -148,17 +148,17 @@ def test_ordering_for_realization_coordinate(self):
+    def test_ordering_for_percentile_coordinate(self):
         """Test that the cube has been reordered, if it is originally in an
-        undesirable order and the cube contains a "percentile_over"
+        undesirable order and the cube contains a "percentile"
```
Nit-picking but can you remove the line break here please.

Katie-Howard replied:
Done.

Katie-Howard commented:
I think I have replied/changed everything that has been commented on. I know that @cgsandford would like to remove the percentile coord option from lib/improver/tests/set_up_test_cubes.py but we have agreed to not put that in this PR. I am happy to do it in a different PR on this ticket, or we can make a new ticket, whatever is best.

fionaRust commented on lib/improver/ensemble_copula_coupling/ensemble_copula_coupling_utilities.py:
```diff
@@ -161,7 +161,7 @@ def create_cube_with_percentiles(percentiles, template_cube, cube_data,
 
     """
-    percentile_coord_name = custom_name or 'percentile_over_realization'
+    percentile_coord_name = custom_name or 'percentile'
```
I'm sorry, I've just noticed "custom_name" isn't in the docstring. I know it isn't your change but would you mind fixing it? Also cube_units is also missing from the docstring

fionaRust replied:
Thanks for fixing this

cgsandford commented on lib/improver/ensemble_copula_coupling/ensemble_copula_coupling_utilities.py:
```diff
@@ -161,11 +163,9 @@ def create_cube_with_percentiles(percentiles, template_cube, cube_data,
-        np.float32(percentiles), long_name=percentile_coord_name,
-        units=unit.Unit("%"), var_name=percentile_coord_name)
+        np.float32(percentiles), long_name='percentile',
+        units=unit.Unit("%"), var_name='percentile')
```
You can also remove "custom_name" as an argument to this function...

Katie-Howard replied:
I thought I had, but I'll do that now.
